quickwit-storage/src/split_creation.rs

```diff
@@ -0,0 +1,404 @@
+    /// Files and their target range in the split bundle.
+    pub files_and_offsets: Vec<(PathBuf, Range<usize>)>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<usize>,
+    /// data of the footer, containng
+    /// [bundle metadata, bundle metadata len 8byte, hotcache, hotcache len 8 bytes]
+    pub footer_bytes: Vec<u8>,
+    /// The end offset of the data provider [File, .. File, SplitFooter]
+    pub block_offsets: Vec<Range<usize>>,
+}
+
+impl SplitPayload {
+    /// Get start and end block index for a range, where start of the Range is in the start block
+    /// and end of the range is in the end block.
+    fn get_start_and_end_block_index(&self, range: Range<usize>) -> io::Result<(usize, usize)> {
+        get_start_and_end_block_index(&self.block_offsets, range)
+    }
+
+    /// Get datasource for data block at index.
+    ///
+    /// We apply seek now because the returned DataSource trait does not allow it.
+    async fn get_datasource(
+        &self,
+        index: usize,
+        start_file_offset: u64,
+    ) -> io::Result<Box<dyn DataSource>> {
+        if index < self.files_and_offsets.len() {
+            let (path, _) = &self.files_and_offsets[index];
+            let mut ds = Box::new(tokio::fs::File::open(path).await?);
+            if start_file_offset != 0 {
+                ds.seek(SeekFrom::Start(start_file_offset)).await?;
+            }
+            Ok(ds)
+        } else {
+            let mut ds = Box::new(Cursor::new(self.footer_bytes.to_vec()));
+            if start_file_offset != 0 {
+                ds.seek(SeekFrom::Start(start_file_offset)).await?;
+            }
+            Ok(ds)
+        }
+    }
+}
+
+#[async_trait]
+impl PutPayload for SplitPayload {
+    async fn len(&self) -> io::Result<u64> {
+        Ok(self.split_size as u64)
+    }
+
+    async fn range_byte_stream(&self, range: Range<u64>) -> io::Result<ByteStream> {
+        let (start_index, end_index) =
+            self.get_start_and_end_block_index(range.start as usize..range.end as usize)?;
+
+        let start_file_offset = range.start - self.block_offsets[start_index].start as u64;
+        // If we have skip and take in the same block, we need to take the skip into account for
+        // the take calculation
+        let same_block_skip = if start_index == end_index {
+            start_file_offset
+        } else {
+            0
+        };
+        let last_block_take =
+            range.end - self.block_offsets[end_index].start as u64 - same_block_skip;
+
+        // build data_sources, apply seek in first
+        let mut datasources: Vec<Box<dyn DataSource>> = vec![];
+        for index in start_index..=end_index {
+            // seek in first datasource
+            let seek_offset = if index == start_index {
+                start_file_offset
+            } else {
+                0
+            };
+            let mut ds = self.get_datasource(index, seek_offset).await?;
```

Comment:
I think it would have been even simpler by pushing more of the logic in `get_datasource`.

```get_datasource(file_index: usize, range: Range<u64>) -> Box<dyn DataSource>```

Reply:
hmm, I don't think this makes sense a `range` can cover multiple datasources